cpp/src/parquet/row_range.h

```diff
@@ -0,0 +1,87 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <variant>
+#include <vector>
+
+#include "parquet/platform.h"
+
+namespace parquet {
+
+/// RowRanges is a collection of non-overlapping and ascendingly ordered row ranges.
+class PARQUET_EXPORT RowRanges {
+ public:
+  /// \brief A range of contiguous rows represented by an interval.
+  struct IntervalRange {
+    /// Start row of the range (inclusive).
+    int64_t start;
+    /// End row of the range (inclusive).
+    int64_t end;
+  };
+
+  /// \brief A range of contiguous rows represented by a bitmap.
+  struct BitmapRange {
+    /// Start row of the range (inclusive).
+    int64_t offset;
+    /// Zero appended if there are less than 64 elements.
+    uint64_t bitmap;
+  };
+
+  /// \brief An end marker for the row range iterator.
+  struct End {};
+
+  /// \brief An iterator for accessing row ranges in order.
+  class Iterator {
```

Comment:
We have iteration facilities in `arrow/util/iterator.h`, should they be reused instead of defining our own here?